src/hotspot/share/gc/shared/workgroup.cpp

```diff
@@ -352,26 +352,19 @@ void WorkGangBarrierSync::abort() {
 // SubTasksDone functions.
 
 SubTasksDone::SubTasksDone(uint n) :
-  _tasks(NULL), _n_tasks(n), _threads_completed(0) {
+  _tasks(NULL), _n_tasks(n) {
   _tasks = NEW_C_HEAP_ARRAY(bool, n, mtInternal);
-  clear();
-}
-
-bool SubTasksDone::valid() {
-  return _tasks != NULL;
-}
-
-void SubTasksDone::clear() {
   for (uint i = 0; i < _n_tasks; i++) {
     _tasks[i] = false;
   }
-  _threads_completed = 0;
 }
 
-void SubTasksDone::all_tasks_completed_impl(uint n_threads,
-                                            uint skipped[],
-                                            size_t skipped_size) {
 #ifdef ASSERT
+void SubTasksDone::all_tasks_completed_impl(uint skipped[], size_t skipped_size) {
+  if (Atomic::cmpxchg(&_verification_done, false, true)) {
```

Comment:
This verification done check prevents detection of certain kinds of mistakes.  For example, if the first thread did not claim a skipped task but a later one did, we'll miss that.  Given that this function only does anything in a debug-build, and is usually pretty fast because the number of subtasks is small, I don't think there's a good reason to "optimize" it this way.  (Assuming it even is an optimization, as a CAS operation may be relatively expensive.)

Reply:
It's not an optimization; it's for avoiding getting duplicate assertion failures. Considering all uses of `SubTasksDone` are based on task-based closure, all threads will run the identical closure. Therefore, it's unlikely that a subtask is skipped in one thread but claimed in another. Revised the comments to make the intention clearer.